vpr/src/route/rr_graph.cpp

```diff
@@ -2654,94 +2654,103 @@ static void build_unidir_rr_opins(const int i, const int j, const e_side side, c
+        clb_to_clb_directs[i].from_clb_pin_end_index = tile_port.absolute_first_pin_index + end_pin_index;
 
         // Load to pins
         // Parse out the pb_type name, port name, and pin range
-        parse_direct_pin_name(directs[i].to_pin, directs[i].line, &start_pin_index, &end_pin_index, pb_type_name, port_name);
+        parse_direct_pin_name(directs[i].to_pin, directs[i].line, &start_pin_index, &end_pin_index, tile_name, port_name);
 
         // Figure out which type, port, and pin is used
-        for (itype = 0; itype < device_ctx.logical_block_types.size(); ++itype) {
-            if (strcmp(device_ctx.logical_block_types[itype].name, pb_type_name) == 0) {
+        for (const auto& type : device_ctx.physical_tile_types) {
+            if (strcmp(type.name, tile_name) == 0) {
+                physical_tile = &type;
                 break;
             }
         }
 
-        if (itype >= device_ctx.logical_block_types.size()) {
-            vpr_throw(VPR_ERROR_ARCH, get_arch_file_name(), directs[i].line, "Unable to find block %s.\n", pb_type_name);
+        if (physical_tile == nullptr) {
+            VPR_THROW(VPR_ERROR_ARCH, "Unable to find block %s.\n", tile_name);
         }
 
-        clb_to_clb_directs[i].to_clb_type = &device_ctx.logical_block_types[itype];
-        pb_type = clb_to_clb_directs[i].to_clb_type->pb_type;
+        clb_to_clb_directs[i].to_clb_type = physical_tile;
 
-        for (j = 0; j < pb_type->num_ports; j++) {
-            if (strcmp(pb_type->ports[j].name, port_name) == 0) {
+        port_found = false;
+        for (const auto& port : physical_tile->ports) {
+            if (0 == strcmp(port.name, port_name)) {
+                tile_port = port;
+                port_found = true;
                 break;
             }
         }
-        if (j >= pb_type->num_ports) {
-            vpr_throw(VPR_ERROR_ARCH, get_arch_file_name(), directs[i].line, "Unable to find port %s (on block %s).\n", port_name, pb_type_name);
+
+        if (!port_found) {
+            VPR_THROW(VPR_ERROR_ARCH, "Unable to find port %s (on block %s).\n", port_name, tile_name);
         }
 
         if (start_pin_index == OPEN) {
             VTR_ASSERT(start_pin_index == end_pin_index);
             start_pin_index = 0;
-            end_pin_index = pb_type->ports[j].num_pins - 1;
+            end_pin_index = tile_port.num_pins - 1;
         }
 
-        get_blk_pin_from_port_pin(clb_to_clb_directs[i].to_clb_type->index, j, start_pin_index, &clb_to_clb_directs[i].to_clb_pin_start_index);
-        get_blk_pin_from_port_pin(clb_to_clb_directs[i].to_clb_type->index, j, end_pin_index, &clb_to_clb_directs[i].to_clb_pin_end_index);
+        clb_to_clb_directs[i].to_clb_pin_start_index = tile_port.absolute_first_pin_index + start_pin_index;
+        clb_to_clb_directs[i].to_clb_pin_end_index = tile_port.absolute_first_pin_index + end_pin_index;
```

Comment:
IMO it would be better to leave this as it is as this operation is applied only here (this one modifies the `to_clb_pin` indices, while the one above modifies the `from_clb_pin` indices).